src/servers/src/http/handler.rs

```diff
@@ -79,20 +86,95 @@ pub async fn sql(
         .start_timer();
 
     let resp = if let Some(sql) = &sql {
-        if let Some(resp) = validate_schema(sql_handler.clone(), query_ctx.clone(), format).await {
-            return Json(resp);
+        if let Some((status, msg)) = validate_schema(sql_handler.clone(), query_ctx.clone()).await {
+            Err((status, msg))
+        } else {
+            Ok(sql_handler.do_query(sql, query_ctx).await)
         }
-
-        JsonResponse::from_output(sql_handler.do_query(sql, query_ctx).await, format, epoch).await
     } else {
-        JsonResponse::with_error_message(
-            "sql parameter is required.".to_string(),
+        Err((
             StatusCode::InvalidArguments,
-            format,
-        )
+            "sql parameter is required.".to_string(),
+        ))
     };
 
-    Json(resp.with_execution_time(start.elapsed().as_millis()))
+    match resp {
```

Comment:
This code goes very dirty because we flatten errors and execution time into `JsonResponse`.

I have an idea to handle `Result<query::Output, ErrorExt>` at this level and factor out the formatting logic from `GreptimedbV1Response` and `InfluxdbV1Response`.

But it may cause some breaking changes :-(

Reply:
That is, for anywhere we handle `JsonResponse` now, instead we handle `Result<query::Output, ErrorExt>`. And returns metadata in headers.

This helps also we can return a proper HTTP error status code instead of the current Status Code = 200 and error in the output.

Reply:
For testing, we may change JsonResponse to a SqlResponse enum:

```rust
enum SqlResponse {
  GreptimedbV1(...),
  InfluxdbV1(...),
  Csv(...),
}
```

and implement IntoResponse on it.

These further code refactoring is trivial to do, but we should first align on the manner. That is, we currently return always StatusCode = 200 in HTTP response and flatten Metadata in JSON reposonse.

The first one is suboptimal and even incompatible with InfluxDB's output since they returns StatusCode != 200:

```
$ curl -i -XPOST 'http://localhost:8086/query?u=myusername&p=notmypassword' --data-urlencode 'q=CREATE DATABASE "mydb"'

HTTP/1.1 401 Unauthorized
Content-Type: application/json; charset=utf-8
X-Influxdb-Build: OSS
X-Influxdb-Version: 2.7.3
X-Platform-Error-Code: unauthorized
Date: Mon, 01 Jan 2024 01:58:56 GMT
Content-Length: 48

{"code":"unauthorized","message":"Unauthorized"}
```

The second one prevent we returns bare CSV output in the payload.

Reply:
It seems to me the problem roots in the situation that format=csv actually has two output, csv for query result and json for error. Adding an http header to indicate the output format is LGTM.

Reply:
OK. I'm going to do a refactor to have

```rust
enum SqlResponse {
  GreptimedbV1(...),
  InfluxdbV1(...),
  Csv(...),
  Error(...),
}
```

and change the related codepath in this week.

This enum will implement "IntoResponse" to adapt HTTP response, and generated from `Result<query::Output, ErrorExt>` so that our codebase is more orthogonal.